src/indexing.jl

```diff
@@ -0,0 +1,502 @@
+is returned, indicating that a new array needs to be reconstructed. This method permits
+customizing the terimnal behavior of the indexing pipeline based on arguments passed
+to `ArrayInterface.getindex`
+"""
+abstract type UnsafeIndex <: Function end
+
+struct UnsafeElement <: UnsafeIndex end
+const unsafe_element = UnsafeElement()
+
+struct UnsafeCollection <: UnsafeIndex end
+const unsafe_collection = UnsafeCollection()
+
+# 1-arg
+UnsafeIndex(x) = UnsafeIndex(typeof(x))
+UnsafeIndex(x::UnsafeIndex) = x
+UnsafeIndex(::Type{T}) where {T<:Integer} = unsafe_element
+UnsafeIndex(::Type{T}) where {T<:AbstractArray} = unsafe_collection
+
+# 2-arg
+UnsafeIndex(x::UnsafeIndex, y::UnsafeElement) = x
+UnsafeIndex(x::UnsafeElement, y::UnsafeIndex) = y
+UnsafeIndex(x::UnsafeElement, y::UnsafeElement) = x
+UnsafeIndex(x::UnsafeCollection, y::UnsafeCollection) = x
+
+
+# tuple
+UnsafeIndex(x::Tuple{I}) where {I} = UnsafeIndex(I)
+@inline function UnsafeIndex(x::Tuple{I,Vararg{Any}}) where {I}
+    return UnsafeIndex(UnsafeIndex(I), UnsafeIndex(tail(x)))
+end
+
+"""
+    unsafe_getindex(A, inds)
+
+Indexes into `A` given `inds`. This method assumes that `inds` have already been
+bounds checked.
+"""
+unsafe_getindex(A, inds) = unsafe_getindex(UnsafeIndex(inds), A, inds)
+unsafe_getindex(::UnsafeElement, A, inds) = unsafe_get_element(A, inds)
+unsafe_getindex(::UnsafeCollection, A, inds) = unsafe_get_collection(A, inds)
+
+"""
+    unsafe_get_element(A::AbstractArray{T}, inds::Tuple) -> T
+
+Returns an element of `A` at the indices `inds`. This method assumes all `inds`
+have been checked for being inbounds. Any new array type using `ArrayInterface.getindex`
+must define `unsafe_get_element(::NewArrayType, inds)`
+"""
+function unsafe_get_element(A, inds)
+    throw(MethodError(unsafe_getindex, (A, inds)))
+end
+function unsafe_get_element(A::Array, inds)
+    if inds isa Tuple{Vararg{Int}}
+        return Base.arrayref(false, A, inds...)
+    else
+        throw(MethodError(unsafe_get_element, (A, inds)))
+    end
+end
+@inline function unsafe_get_element(A::LinearIndices, inds)
+    return Int(Base._to_linear_index(A, inds...))
+end
+@inline function unsafe_get_element(A::CartesianIndices, inds)
+    return CartesianIndex(Base._to_subscript_indices(A, inds...))
+end
+
+# This is based on Base._unsafe_getindex from https://github.com/JuliaLang/julia/blob/c5ede45829bf8eb09f2145bfd6f089459d77b2b1/base/multidimensional.jl#L755
+"""
+    unsafe_get_collection(A, inds)
+
+Returns a collection of `A` given `inds`. `inds` is assumed to be bounds checked prior.
+"""
+function unsafe_get_collection(A, inds)
+    axs = to_axes(A, inds)
+    dest = similar(A, axs)
+    map(Base.unsafe_length, axes(dest)) == map(Base.unsafe_length, axs) || throw_checksize_error(dest, axs)
+    Base._unsafe_getindex!(dest, A, inds...) # usually a generated function, don't allow it to impact inference result
+    return dest
+end
```

Comment:
Should this be replaced by:

```julia
unsafe_get_collection(A, inds) = unsafe_get_collection(device(A), A, inds)
function unsafe_get_collection(::CPUIndex, A, inds)
    axs = to_axes(A, inds)
    dest = similar(A, axs)
    map(Base.unsafe_length, axes(dest)) == map(Base.unsafe_length, axs) || Base.throw_checksize_error(dest, axs)
    Base._unsafe_getindex!(dest, A, inds...) # usually a generated function, don't allow it to impact inference result
    return dest
end

function unsafe_get_collection(::CPUPointer, A, inds)
    data = vload(pointer(A), unroll(A, inds))
    return unsafe_reconstruct(A, data; axes=to_axes(A, inds))
end
```

This would require formalizing some of the interface for raw access to memory, but it would pretty much make this complete.